data-serving/data-service/src/controllers/case.ts

```diff
@@ -301,41 +298,35 @@ export class CasesController {
                 });
                 return;
             }
-            const bulkWriteResult = await Case.bulkWrite(
-                // eslint-disable-next-line @typescript-eslint/no-explicit-any
-                req.body.cases.map((c: any) => {
-                    if (
-                        c.caseReference?.sourceId &&
-                        c.caseReference?.sourceEntryId
-                    ) {
-                        return {
-                            updateOne: {
-                                filter: {
-                                    'caseReference.sourceId':
-                                        c.caseReference.sourceId,
-                                    'caseReference.sourceEntryId':
-                                        c.caseReference.sourceEntryId,
-                                },
-                                update: c,
-                                upsert: true,
-                            },
-                        };
-                    } else {
-                        return {
-                            insertOne: {
-                                document: c,
-                            },
-                        };
-                    }
-                }),
-                { ordered: false },
-            );
+            // Use this method rather than Case.bulkWrite() as that method
+            // causes an out of memory error for a large number of cases.
+            const bulk = Case.collection.initializeUnorderedBulkOp();
+            // eslint-disable-next-line @typescript-eslint/no-explicit-any
+            req.body.cases.map((c: any) => {
+                delete c.caseCount;
+                if (
+                    c.caseReference?.sourceId &&
+                    c.caseReference?.sourceEntryId
+                ) {
+                    delete c._id;
+                    bulk.find({
+                        'caseReference.sourceId': c.caseReference.sourceId,
+                        'caseReference.sourceEntryId':
+                            c.caseReference.sourceEntryId,
+                    })
+                        .upsert()
+                        .updateOne({ $set: c });
+                } else {
+                    bulk.insert(c);
+                }
+            });
+            const bulkWriteResult = await bulk.execute();
             res.status(200).json({
                 phase: 'UPSERT',
                 numCreated:
-                    (bulkWriteResult.insertedCount || 0) +
-                    (bulkWriteResult.upsertedCount || 0),
-                numUpdated: bulkWriteResult.modifiedCount,
+                    (bulkWriteResult.nInserted || 0) +
```

Comment:
nit but I really prefer `numInserted` instead of `nInserted`
(yes it comes from a gopher I know but this is not go... ;) )

Reply:
It's a field on the bulkWriteResult that we can't change :/ it is strange that they change from numInserted to nInserted depending on which bulk method is used.